File: backend/app/schemas/nomina.py
"""
Schemas del módulo de nómina (fase inicial).
"""
from datetime import date, datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class NominaEmpleadoCreate(BaseModel):
    sucursal_id: Optional[UUID] = None
    centro_costo_id: Optional[UUID] = None
    codigo_interno: Optional[str] = Field(default=None, max_length=50)
    documento_tipo: str = Field(min_length=2, max_length=20)
    documento_numero: str = Field(min_length=4, max_length=40)
    nombres: str = Field(min_length=2, max_length=120)
    apellidos: str = Field(min_length=2, max_length=120)
    email: Optional[str] = Field(default=None, max_length=255)
    celular: Optional[str] = Field(default=None, max_length=30)
    fecha_ingreso: date


class NominaEmpleadoResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    sucursal_id: Optional[UUID] = None
    centro_costo_id: Optional[UUID] = None
    codigo_interno: Optional[str] = None
    documento_tipo: str
    documento_numero: str
    nombres: str
    apellidos: str
    email: Optional[str] = None
    celular: Optional[str] = None
    fecha_ingreso: date
    activo: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class NominaContratoCreate(BaseModel):
    empleado_id: UUID
    centro_costo_id: Optional[UUID] = None
    es_salario_integral: bool = False
    tipo_contrato: str
    periodicidad: str
    salario_base: Decimal = Field(gt=0)
    fecha_inicio: date
    fecha_fin: Optional[date] = None
    observaciones: Optional[str] = None


class NominaContratoResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    empleado_id: UUID
    centro_costo_id: Optional[UUID] = None
    es_salario_integral: bool
    tipo_contrato: str
    periodicidad: str
    salario_base: Decimal
    fecha_inicio: date
    fecha_fin: Optional[date] = None
    estado: str
    observaciones: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class NominaPeriodoCreate(BaseModel):
    anio: str = Field(min_length=4, max_length=4)
    mes: str = Field(min_length=2, max_length=2)
    fecha_inicio: date
    fecha_fin: date
    fecha_pago: Optional[date] = None
    observaciones: Optional[str] = None


class NominaPeriodoEstadoUpdate(BaseModel):
    estado: str
    observaciones: Optional[str] = None


class NominaPeriodoResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    anio: str
    mes: str
    fecha_inicio: date
    fecha_fin: date
    fecha_pago: Optional[date] = None
    estado: str
    observaciones: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class NominaNovedadCreate(BaseModel):
    periodo_id: UUID
    empleado_id: UUID
    tipo: str
    concepto: str = Field(min_length=2, max_length=120)
    unidades: Decimal = Field(default=Decimal("1"), ge=0)
    valor_unitario: Decimal = Field(default=Decimal("0"), ge=0)
    valor_total: Decimal = Field(default=Decimal("0"), ge=0)
    observaciones: Optional[str] = None


class NominaNovedadResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    periodo_id: UUID
    empleado_id: UUID
    tipo: str
    concepto: str
    unidades: Decimal
    valor_unitario: Decimal
    valor_total: Decimal
    observaciones: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class NominaLiquidacionResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    periodo_id: UUID
    empleado_id: UUID
    contrato_id: UUID
    salario_base: Decimal
    total_devengos: Decimal
    total_deducciones: Decimal
    neto_pagar: Decimal
    auxilio_transporte_devengo: Decimal
    base_cotizacion: Decimal
    aporte_salud_empleado: Decimal
    aporte_pension_empleado: Decimal
    aporte_fsp_empleado: Decimal
    aporte_subsistencia_empleado: Decimal
    retencion_fuente_empleado: Decimal
    aporte_salud_empresa: Decimal
    aporte_pension_empresa: Decimal
    aporte_arl_empresa: Decimal
    aporte_caja_empresa: Decimal
    aporte_sena_empresa: Decimal
    aporte_icbf_empresa: Decimal
    provision_prima: Decimal
    provision_cesantias: Decimal
    provision_intereses_cesantias: Decimal
    provision_vacaciones: Decimal
    costo_total_empresa: Decimal
    desprendible_folio: Optional[str] = None
    desprendible_version: int
    desprendible_pdf_sha256: Optional[str] = None
    desprendible_generated_at: Optional[datetime] = None
    observaciones: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class NominaPreliquidacionResumen(BaseModel):
    periodo_id: UUID
    empleados_liquidados: int
    total_salario_base: Decimal
    total_devengos: Decimal
    total_deducciones: Decimal
    total_neto_pagar: Decimal


class NominaConceptoDetalle(BaseModel):
    tipo: str
    concepto: str
    unidades: Decimal
    valor_unitario: Decimal
    valor_total: Decimal


class NominaDesprendibleBaseResponse(BaseModel):
    tenant_id: UUID
    periodo_id: UUID
    liquidacion_id: UUID
    empleado_id: UUID
    empleado_nombre: str
    empleado_documento: str
    salario_base: Decimal
    total_devengos: Decimal
    total_deducciones: Decimal
    neto_pagar: Decimal
    devengos: list[NominaConceptoDetalle]
    deducciones: list[NominaConceptoDetalle]


class NominaDesprendibleMetaResponse(BaseModel):
    liquidacion_id: UUID
    periodo_id: UUID
    empleado_id: UUID
    folio: Optional[str] = None
    version: int
    pdf_relpath: Optional[str] = None
    pdf_sha256: Optional[str] = None
    generated_at: Optional[datetime] = None


class NominaDesprendibleVersionResponse(BaseModel):
    id: UUID
    liquidacion_id: UUID
    periodo_id: UUID
    empleado_id: UUID
    folio: Optional[str] = None
    version: int
    pdf_relpath: str
    pdf_sha256: str
    generated_at: datetime
    generated_by: UUID
    motivo: str

    model_config = ConfigDict(from_attributes=True)


class NominaCentroCostoCreate(BaseModel):
    sucursal_id: Optional[UUID] = None
    codigo: str = Field(min_length=1, max_length=30)
    nombre: str = Field(min_length=2, max_length=160)
    descripcion: Optional[str] = None


class NominaCentroCostoResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    sucursal_id: Optional[UUID] = None
    codigo: str
    nombre: str
    descripcion: Optional[str] = None
    activo: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class NominaParametroLegalResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    salario_minimo_mensual: Decimal
    auxilio_transporte_mensual: Decimal
    uvt: Decimal
    tope_ibc_smmlv: Decimal
    umbral_exoneracion_smmlv: Decimal
    exoneracion_aportes_activa: bool
    aplica_auxilio_transporte: bool
    umbral_auxilio_transporte_smmlv: Decimal
    aplica_fsp: bool
    umbral_fsp_smmlv: Decimal
    pct_fsp_base: Decimal
    aplica_subsistencia: bool
    aplica_retencion_fuente: bool
    umbral_retencion_uvt: Decimal
    pct_retencion_base: Decimal
    pct_ibc_salario_integral: Decimal
    pct_salud_empleado: Decimal
    pct_pension_empleado: Decimal
    pct_salud_empresa: Decimal
    pct_pension_empresa: Decimal
    pct_arl_empresa: Decimal
    pct_caja_empresa: Decimal
    pct_sena_empresa: Decimal
    pct_icbf_empresa: Decimal
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class NominaParametroLegalUpdate(BaseModel):
    salario_minimo_mensual: Optional[Decimal] = Field(default=None, ge=0)
    auxilio_transporte_mensual: Optional[Decimal] = Field(default=None, ge=0)
    uvt: Optional[Decimal] = Field(default=None, ge=0)
    tope_ibc_smmlv: Optional[Decimal] = Field(default=None, gt=0)
    umbral_exoneracion_smmlv: Optional[Decimal] = Field(default=None, gt=0)
    exoneracion_aportes_activa: Optional[bool] = None
    aplica_auxilio_transporte: Optional[bool] = None
    umbral_auxilio_transporte_smmlv: Optional[Decimal] = Field(default=None, gt=0)
    aplica_fsp: Optional[bool] = None
    umbral_fsp_smmlv: Optional[Decimal] = Field(default=None, gt=0)
    pct_fsp_base: Optional[Decimal] = Field(default=None, ge=0, le=1)
    aplica_subsistencia: Optional[bool] = None
    aplica_retencion_fuente: Optional[bool] = None
    umbral_retencion_uvt: Optional[Decimal] = Field(default=None, ge=0)
    pct_retencion_base: Optional[Decimal] = Field(default=None, ge=0, le=1)
    pct_ibc_salario_integral: Optional[Decimal] = Field(default=None, ge=0, le=1)
    pct_salud_empleado: Optional[Decimal] = Field(default=None, ge=0, le=1)
    pct_pension_empleado: Optional[Decimal] = Field(default=None, ge=0, le=1)
    pct_salud_empresa: Optional[Decimal] = Field(default=None, ge=0, le=1)
    pct_pension_empresa: Optional[Decimal] = Field(default=None, ge=0, le=1)
    pct_arl_empresa: Optional[Decimal] = Field(default=None, ge=0, le=1)
    pct_caja_empresa: Optional[Decimal] = Field(default=None, ge=0, le=1)
    pct_sena_empresa: Optional[Decimal] = Field(default=None, ge=0, le=1)
    pct_icbf_empresa: Optional[Decimal] = Field(default=None, ge=0, le=1)

File: frontend/src/api/nomina.ts
import apiClient, { apiBaseUrl } from './client';

export interface NominaPeriodo {
  id: string;
  tenant_id: string;
  anio: string;
  mes: string;
  fecha_inicio: string;
  fecha_fin: string;
  fecha_pago?: string | null;
  estado: string;
  observaciones?: string | null;
  created_at: string;
}

export interface NominaPeriodoCreatePayload {
  anio: string;
  mes: string;
  fecha_inicio: string;
  fecha_fin: string;
  fecha_pago?: string | null;
  observaciones?: string | null;
}

export interface NominaLiquidacion {
  id: string;
  tenant_id: string;
  periodo_id: string;
  empleado_id: string;
  contrato_id: string;
  salario_base: number;
  total_devengos: number;
  total_deducciones: number;
  neto_pagar: number;
  auxilio_transporte_devengo: number;
  base_cotizacion: number;
  aporte_salud_empleado: number;
  aporte_pension_empleado: number;
  aporte_fsp_empleado: number;
  aporte_subsistencia_empleado: number;
  retencion_fuente_empleado: number;
  aporte_salud_empresa: number;
  aporte_pension_empresa: number;
  aporte_arl_empresa: number;
  aporte_caja_empresa: number;
  aporte_sena_empresa: number;
  aporte_icbf_empresa: number;
  provision_prima: number;
  provision_cesantias: number;
  provision_intereses_cesantias: number;
  provision_vacaciones: number;
  costo_total_empresa: number;
  desprendible_folio?: string | null;
  desprendible_version: number;
  desprendible_pdf_sha256?: string | null;
  desprendible_generated_at?: string | null;
  observaciones?: string | null;
  created_at: string;
  updated_at: string;
}

export interface NominaEmpleado {
  id: string;
  tenant_id: string;
  sucursal_id?: string | null;
  centro_costo_id?: string | null;
  codigo_interno?: string | null;
  documento_tipo: string;
  documento_numero: string;
  nombres: string;
  apellidos: string;
  email?: string | null;
  celular?: string | null;
  fecha_ingreso?: string;
  activo?: string;
  created_at?: string;
}

export interface NominaEmpleadoCreatePayload {
  sucursal_id?: string | null;
  centro_costo_id?: string | null;
  codigo_interno?: string | null;
  documento_tipo: string;
  documento_numero: string;
  nombres: string;
  apellidos: string;
  email?: string | null;
  celular?: string | null;
  fecha_ingreso: string;
}

export interface NominaContrato {
  id: string;
  tenant_id: string;
  empleado_id: string;
  es_salario_integral: boolean;
  centro_costo_id?: string | null;
  tipo_contrato: string;
  periodicidad: string;
  salario_base: number;
  fecha_inicio: string;
  fecha_fin?: string | null;
  estado: string;
  observaciones?: string | null;
  created_at: string;
}

export interface NominaContratoCreatePayload {
  empleado_id: string;
  es_salario_integral?: boolean;
  centro_costo_id?: string | null;
  tipo_contrato: 'fijo' | 'indefinido' | 'obra_labor' | 'aprendizaje' | 'temporal';
  periodicidad: 'quincenal' | 'mensual';
  salario_base: number;
  fecha_inicio: string;
  fecha_fin?: string | null;
  observaciones?: string | null;
}

export interface NominaNovedad {
  id: string;
  tenant_id: string;
  periodo_id: string;
  empleado_id: string;
  tipo: 'devengo' | 'deduccion';
  concepto: string;
  unidades: number;
  valor_unitario: number;
  valor_total: number;
  observaciones?: string | null;
  created_at: string;
}

export interface NominaNovedadCreatePayload {
  periodo_id: string;
  empleado_id: string;
  tipo: 'devengo' | 'deduccion';
  concepto: string;
  unidades: number;
  valor_unitario: number;
  valor_total: number;
  observaciones?: string | null;
}

export interface NominaPreliquidacionResumen {
  periodo_id: string;
  empleados_liquidados: number;
  total_salario_base: number;
  total_devengos: number;
  total_deducciones: number;
  total_neto_pagar: number;
}

export interface NominaParametroLegal {
  id: string;
  tenant_id: string;
  salario_minimo_mensual: number;
  auxilio_transporte_mensual: number;
  uvt: number;
  tope_ibc_smmlv: number;
  umbral_exoneracion_smmlv: number;
  exoneracion_aportes_activa: boolean;
  aplica_auxilio_transporte: boolean;
  umbral_auxilio_transporte_smmlv: number;
  aplica_fsp: boolean;
  umbral_fsp_smmlv: number;
  pct_fsp_base: number;
  aplica_subsistencia: boolean;
  aplica_retencion_fuente: boolean;
  umbral_retencion_uvt: number;
  pct_retencion_base: number;
  pct_ibc_salario_integral: number;
  pct_salud_empleado: number;
  pct_pension_empleado: number;
  pct_salud_empresa: number;
  pct_pension_empresa: number;
  pct_arl_empresa: number;
  pct_caja_empresa: number;
  pct_sena_empresa: number;
  pct_icbf_empresa: number;
  updated_at?: string | null;
}

export type NominaParametroLegalUpdatePayload = Partial<Omit<NominaParametroLegal, 'id' | 'tenant_id' | 'updated_at'>>;

export interface NominaCentroCosto {
  id: string;
  tenant_id: string;
  sucursal_id?: string | null;
  codigo: string;
  nombre: string;
  descripcion?: string | null;
  activo: string;
  created_at: string;
}

export interface NominaCentroCostoCreatePayload {
  sucursal_id?: string | null;
  codigo: string;
  nombre: string;
  descripcion?: string | null;
}

export interface NominaDesprendibleVersion {
  id: string;
  liquidacion_id: string;
  periodo_id: string;
  empleado_id: string;
  folio?: string | null;
  version: number;
  pdf_relpath: string;
  pdf_sha256: string;
  generated_at: string;
  generated_by: string;
  motivo: string;
}

export interface NominaReemitirResponse {
  ok: boolean;
  liquidacion_id: string;
  folio: string;
  version: number;
  pdf_sha256?: string | null;
  pdf_bytes: number;
}

async function fetchPdf(path: string): Promise<{ blob: Blob; filename: string }> {
  const token = localStorage.getItem('access_token');
  if (!token) {
    throw new Error('No hay token de autenticacion');
  }
  const response = await fetch(`${apiBaseUrl}${path}`, {
    method: 'GET',
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!response.ok) {
    throw new Error(`No se pudo descargar el PDF (HTTP ${response.status})`);
  }
  const blob = await response.blob();
  const contentDisposition = response.headers.get('content-disposition') || '';
  const match = contentDisposition.match(/filename="?(.+?)"?$/i);
  return {
    blob,
    filename: (match?.[1] || 'desprendible.pdf').trim(),
  };
}

export const nominaApi = {
  listarPeriodos: async (): Promise<NominaPeriodo[]> => {
    const response = await apiClient.get<NominaPeriodo[]>('/nomina/periodos');
    return response.data;
  },

  crearPeriodo: async (payload: NominaPeriodoCreatePayload): Promise<NominaPeriodo> => {
    const response = await apiClient.post<NominaPeriodo>('/nomina/periodos', payload);
    return response.data;
  },

  obtenerParametrosLegales: async (): Promise<NominaParametroLegal> => {
    const response = await apiClient.get<NominaParametroLegal>('/nomina/parametros-legales');
    return response.data;
  },

  actualizarParametrosLegales: async (
    payload: NominaParametroLegalUpdatePayload
  ): Promise<NominaParametroLegal> => {
    const response = await apiClient.put<NominaParametroLegal>('/nomina/parametros-legales', payload);
    return response.data;
  },

  listarCentrosCosto: async (): Promise<NominaCentroCosto[]> => {
    const response = await apiClient.get<NominaCentroCosto[]>('/nomina/centros-costo');
    return response.data;
  },

  crearCentroCosto: async (payload: NominaCentroCostoCreatePayload): Promise<NominaCentroCosto> => {
    const response = await apiClient.post<NominaCentroCosto>('/nomina/centros-costo', payload);
    return response.data;
  },

  listarEmpleados: async (): Promise<NominaEmpleado[]> => {
    const response = await apiClient.get<NominaEmpleado[]>('/nomina/empleados');
    return response.data;
  },

  crearEmpleado: async (payload: NominaEmpleadoCreatePayload): Promise<NominaEmpleado> => {
    const response = await apiClient.post<NominaEmpleado>('/nomina/empleados', payload);
    return response.data;
  },

  listarContratos: async (): Promise<NominaContrato[]> => {
    const response = await apiClient.get<NominaContrato[]>('/nomina/contratos');
    return response.data;
  },

  crearContrato: async (payload: NominaContratoCreatePayload): Promise<NominaContrato> => {
    const response = await apiClient.post<NominaContrato>('/nomina/contratos', payload);
    return response.data;
  },

  listarLiquidacionesPeriodo: async (
    periodoId: string,
    params?: { empleado_id?: string; sucursal_id?: string; centro_costo_id?: string }
  ): Promise<NominaLiquidacion[]> => {
    const response = await apiClient.get<NominaLiquidacion[]>(`/nomina/periodos/${periodoId}/liquidaciones`, { params });
    return response.data;
  },

  listarNovedadesPeriodo: async (
    periodoId: string,
    params?: { empleado_id?: string; sucursal_id?: string; centro_costo_id?: string }
  ): Promise<NominaNovedad[]> => {
    const response = await apiClient.get<NominaNovedad[]>(`/nomina/novedades`, {
      params: { periodo_id: periodoId, ...params },
    });
    return response.data;
  },

  crearNovedad: async (payload: NominaNovedadCreatePayload): Promise<NominaNovedad> => {
    const response = await apiClient.post<NominaNovedad>('/nomina/novedades', payload);
    return response.data;
  },

  listarVersionesDesprendible: async (liquidacionId: string): Promise<NominaDesprendibleVersion[]> => {
    const response = await apiClient.get<NominaDesprendibleVersion[]>(
      `/nomina/liquidaciones/${liquidacionId}/desprendibles/versiones`
    );
    return response.data;
  },

  reemitirDesprendible: async (liquidacionId: string): Promise<NominaReemitirResponse> => {
    const response = await apiClient.post<NominaReemitirResponse>(
      `/nomina/liquidaciones/${liquidacionId}/reemitir-desprendible`
    );
    return response.data;
  },

  preliquidarPeriodo: async (periodoId: string): Promise<NominaPreliquidacionResumen> => {
    const response = await apiClient.post<NominaPreliquidacionResumen>(`/nomina/periodos/${periodoId}/preliquidar`);
    return response.data;
  },

  aprobarPeriodo: async (periodoId: string): Promise<NominaPeriodo> => {
    const response = await apiClient.post<NominaPeriodo>(`/nomina/periodos/${periodoId}/aprobar`);
    return response.data;
  },

  cerrarPeriodo: async (periodoId: string): Promise<NominaPeriodo> => {
    const response = await apiClient.post<NominaPeriodo>(`/nomina/periodos/${periodoId}/cerrar`);
    return response.data;
  },

  marcarPagadaPeriodo: async (periodoId: string): Promise<NominaPeriodo> => {
    const response = await apiClient.post<NominaPeriodo>(`/nomina/periodos/${periodoId}/marcar-pagada`);
    return response.data;
  },

  descargarDesprendibleActual: (liquidacionId: string) =>
    fetchPdf(`/nomina/liquidaciones/${liquidacionId}/desprendible.pdf`),

  descargarDesprendibleVersion: (liquidacionId: string, version: number) =>
    fetchPdf(`/nomina/liquidaciones/${liquidacionId}/desprendibles/versiones/${version}/pdf`),
};
